src/client-side-encryption/stateMachine.ts

```diff
@@ -43,90 +53,135 @@ const INSECURE_TLS_OPTIONS = [
+    console.error(msg);
+  }
+}
 
-/**
- * @ignore
- * @callback StateMachine~fetchCollectionInfoCallback
- * @param {Error} [err] If present, indicates that fetching the collection info failed with the given error
- * @param {object} [result] If present, is the fetched collection info for the first collection to match the given filter
- * @returns {void}
- */
+declare module 'mongodb-client-encryption' {
+  // the properties added to `MongoCryptContext` here are only used for the `StateMachine`'s
+  // execute method and are not part of the C++ bindings.
+  interface MongoCryptContext {
+    id: number;
+    document: Document;
+    ns: string;
+  }
+}
 
 /**
- * @ignore
- * @callback StateMachine~markCommandCallback
- * @param {Error} [err] If present, indicates that marking the command failed with the given error
- * @param {Buffer} [result] If present, is the marked command serialized into bson
- * @returns {void}
+ * TLS options to use when connecting. The spec specifically calls out which insecure
+ * tls options are not allowed:
+ *
+ *  - tlsAllowInvalidCertificates
+ *  - tlsAllowInvalidHostnames
+ *  - tlsInsecure
+ *  - tlsDisableOCSPEndpointCheck
+ *  - tlsDisableCertificateRevocationCheck
  */
+export type CSFLETlsOptions = Pick<
+  MongoClientOptions,
+  'tlsCAFile' | 'tlsCertificateKeyFile' | 'tlsCertificateKeyFilePassword'
+>;
+
+export type CSFLEKMSTlsOptions = {
+  aws?: CSFLETlsOptions;
+  gcp?: CSFLETlsOptions;
+  kmip?: CSFLETlsOptions;
+  local?: CSFLETlsOptions;
+  azure?: CSFLETlsOptions;
+};
+
+export interface StateMachineExecutable {
+  _keyVaultNamespace: string;
+  _keyVaultClient: MongoClient;
+  _metaDataClient?: MongoClient;
+  _mongocryptdClient?: MongoClient;
+  _mongocryptdManager?: MongocryptdManager;
+  askForKMSCredentials: () => Promise<KMSProviders | Buffer>;
+}
 
-/**
- * @ignore
- * @callback StateMachine~fetchKeysCallback
- * @param {Error} [err] If present, indicates that fetching the keys failed with the given error
- * @param {object[]} [result] If present, is all the keys from the keyVault collection that matched the given filter
- */
+export type StateMachineOptions = {
+  /** socks5 proxy options, if set. */
+  proxyOptions: ProxyOptions;
+
+  /** TLS options for KMS requests, if set. */
+  tlsOptions: CSFLEKMSTlsOptions;
+} & Pick<BSONSerializeOptions, 'promoteLongs' | 'promoteValues'> &
+  CommandOptions;
 
 /**
- * @ignore
+ * @internal
  * An internal class that executes across a MongoCryptContext until either
  * a finishing state or an error is reached. Do not instantiate directly.
- * @class StateMachine
  */
-class StateMachine {
-  constructor(options) {
-    this.options = options || {};
-
-    this.executeAsync = promisify((autoEncrypter, context, callback) =>
-      this.execute(autoEncrypter, context, callback)
-    );
+export class StateMachine {
+  constructor(
+    private options: StateMachineOptions,
+    private bsonOptions = pluckBSONSerializeOptions(options)
```

Comment:
This is new.  Previously, we simply passed the `options` object into `serialize` to provide bson options.  I figured it makes sense to pull off all supported BSON serialization options and store them separately for clarity.

Reply:
SGTM